### PR TITLE
New exec code path rebased

### PR DESCRIPTION
**Pull request by ryoqun**
this is yet another pr for unified scheduler (prev: #31239)

todo

- [x] rename new crate and publish boilerplate ones.
- [ ] handle replay_apply_costs
- [x] considering early-bail-out even for sending transactions to scheduler in the case of encountering fatal tx errors.
- [x] proper drop impl.
- [ ] fix metrics.
- [ ] track down or triage the commit on frozen bank panic...
- [x] expose `--unified-scheduler-handler-threads=N`
- [ ] expose `--unified-scheduler-individual-transaction-timings`?

### CODE REVIEW

File: scheduler-pool/src/lib.rs [hunk 1]
```diff
@@ -253,8 +253,8 @@ impl<TH: ScheduledTransactionHandler<SEA>, SEA: ScheduleExecutionArg> InstalledS
         self.id
     }
 
-    fn pool(&self) -> InstalledSchedulerPoolArc<SEA> {
-        self.pool.clone()
+    fn return_to_pool(self: Box<Self>) {
+        self.pool.clone().return_scheduler(self)
```
Comment (ryoqun) [inline scheduler-pool/src/lib.rs]: here

File: scheduler-pool/src/lib.rs [hunk 2]
```diff
@@ -110,6 +110,15 @@ impl<
...
     }
+
+    pub fn return_scheduler(&self, scheduler: Box<dyn InstalledScheduler<SEA>>) {
+        assert!(scheduler.context().is_none());
+
+        self.schedulers
+            .lock()
+            .expect("not poisoned")
+            .push(scheduler);
+    }
```
Comment (ryoqun) [inline scheduler-pool/src/lib.rs]: here 2

File: scheduler-pool/src/lib.rs (hunk 1 above)
Comment (ryoqun) [reply scheduler-pool/src/lib.rs]: oh, the `.clone()` can be removed, btw... ;)

Comment (ryoqun) [reply scheduler-pool/src/lib.rs]: well, never mind. i can't. borrow checker would complain...

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -5211,14 +5211,17 @@ impl Bank {
 
         execution_time.stop();
 
-        const SHRINK_LOADED_PROGRAMS_TO_PERCENTAGE: u8 = 90;
-        self.loaded_programs_cache
-            .write()
-            .unwrap()
```
Comment (ryoqun) [inline runtime/src/bank.rs:5217]: here

File: runtime/src/bank_forks.rs [hunk 4]
```diff
@@ -212,21 +220,44 @@ impl BankForks {
...
+        let root_bank = self.root_bank();
+        // drop all non root BankWithScheduler, which causes all schedulers wind down.
+        self.banks.clear();
+        self.uninstall_scheduler_pool();
+        // this cuts circular references of BankForks...
+        root_bank
+            .loaded_programs_cache
+            .write()
+            .unwrap()
+            .unset_fork_graph();
```
Comment (ryoqun) [inline runtime/src/bank_forks.rs:251]: here

File: unified-scheduler-pool/benches/lib.rs
Comment (ryoqun) [inline unified-scheduler-pool/benches/lib.rs:1]: this file!

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

**Pull request closed without merging**